lib/logthrsource/logthrsourcedrv.c

```diff
@@ -0,0 +1,324 @@
+/*
+ * Copyright (c) 2018 Balabit
+ * Copyright (c) 2018 László Várady <laszlo.varady@balabit.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logthrsourcedrv.h"
+#include "mainloop-worker.h"
+#include "messages.h"
+#include "apphook.h"
+
+typedef struct _WakeupCondition
+{
+  GMutex *lock;
+  GCond *cond;
+  gboolean awoken;
+} WakeupCondition;
+
+struct _LogThreadedSourceWorker
+{
```

Comment:
I think the multi-threaded version can be implemented relatively easily on the top of the current implementation. In the case of `LogThreadedFetcherDriver`, it might be useful to run multiple `LogThreadedSourceWorker`s.

Because the API of `LogThreadedSourceDriver` supports external event loops, it is also possible to start an existing server framework or event loop in the dedicated thread that has its own separate worker pool.